### PR TITLE
Implement RFC #774.

### PR DESCRIPTION
**Pull request by ryzhyk**

### CODE REVIEW

File: src/Language/DifferentialDatalog/Parse.hs [hunk 1]
```diff
@@ -408,20 +411,66 @@ apply = Apply nopos <$  reserved "apply" <*> transIdent
...
+    checkNoProg (length args == 2) p
+                $ "The 'group_by' operator must be invoked with two arguments, e.g., 'expr1.group_by(expr2)' or 'group_by(expr1, expr2)'," ++
+                  " but it is invoked with " ++ show (length args)
+    let [project, group_by] = args
+    previous_groupby_clause <- ST.get
+    checkNoProg (isNothing previous_groupby_clause) p
+                $ "'group_by' operator can occur at most once in an expression. Previous occurrence: " ++
+                  show (pos $ rhsGroupBy $ fromJust previous_groupby_clause)
+    ST.put $ Just $ RHSGroupBy "__group" project group_by
+    return $ E $ EVar p "__group"
```
Comment (mbudiu-vmw) [inline src/Language/DifferentialDatalog/Parse.hs:452]: why is this name really fresh?

File: src/Language/DifferentialDatalog/Parse.hs [hunk 2]
```diff
@@ -408,20 +411,66 @@ apply = Apply nopos <$  reserved "apply" <*> transIdent
...
+extractGroupBy e  = return $ E e
+
+-- Deprecated Aggregate syntax.
+aggregate = do
```
Comment (mbudiu-vmw) [inline src/Language/DifferentialDatalog/Parse.hs:456]: can you add a TODO here to add a warning when we have them?

File: src/Language/DifferentialDatalog/Syntax.hs [hunk 3]
```diff
@@ -590,8 +590,32 @@ instance Show Atom where
 -- positive/negative polarity, Boolean conditions, aggregation,
 -- disaggregation (flatmap), inspect operations.
 data RuleRHS = RHSLiteral   {rhsPolarity :: Bool, rhsAtom :: Atom}
-             | RHSCondition {rhsExpr :: Expr}
-             | RHSAggregate {rhsVar :: String, rhsGroupBy :: Expr, rhsAggFunc :: String, rhsAggExpr :: Expr}
+               -- Expression that can filter or map input collation, or both:
+               -- * Filtering: `x == y`
+               -- * Mapping: `var x = f(y)`
+               -- * Filter/map: `Some{x} = f(y)`
+             | RHSCondition {
```
Comment (mbudiu-vmw) [inline src/Language/DifferentialDatalog/Syntax.hs:597]: is this then a good name?

File: test/datalog_tests/group_test.dl [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+    var g = d.group_by(()),
+    Some{var finding} = g.find(|d| d.z.nth(1).unwrap_or_default() == 5).
+
+// filter
+output relation Filter1(val: Vec<Data>)
+
+Filter1(g.filter(|d| d.z.len() == 3)) :-
+    d in Data(),
+    var g = d.group_by(()),
+    d in Data().
```
Comment (mbudiu-vmw) [inline test/datalog_tests/group_test.dl:76]: this is  a weird example.

File: test/datalog_tests/rules.fail.ast.expected [hunk 5]
```diff
@@ -11,24 +11,36 @@ error: ./test/datalog_tests/rules.fail.dl:10:1-11:1: Relation R3 is mutually rec
...
+actual type: ddlog_std::Group<string,string>
+in
+expected type: function( ddlog_std::Group<(string,string),string>):string
+actual type: function( ddlog_std::Group<string,string>):string
 in
-expression '(x, z)'
+expression 'concat_ys'
     var s = Aggregate((x,z), concat_ys(y)).
-                      ^^^^^
+                           ^^^^^^^^^^^
```
Comment (mbudiu-vmw) [inline test/datalog_tests/rules.fail.ast.expected:26]: Do you want to capture the comma in the source position of this expression?

**Review by mbudiu-vmw — APPROVED** [review]

File: lib/ddlog_std.rs [hunk 6]
```diff
@@ -1145,110 +1149,347 @@ pub fn hash128<T: Hash>(x: &T) -> u128 {
...
+ * the DDlog program; hence two groups are iff they have the same
+ * projections. */
+
+impl<K: PartialEq, V: Clone + PartialEq> PartialEq for Group<K, V> {
+    fn eq(&self, other: &Self) -> bool {
+        (self.key_ref() == other.key_ref()) && (self.iter().eq(other.iter()))
+    }
+}
+
+impl<K: PartialEq, V: Clone + PartialEq> Eq for Group<K, V> {}
```
[the remark this answers is not in the document]
Comment (ryzhyk) [reply lib/ddlog_std.rs:1235]: `Eq` does not define any methods.  `==` is part of the `PartialEq` trait.

File: src/Language/DifferentialDatalog/Debug.hs [hunk 7]
```diff
@@ -60,46 +57,29 @@ addBindingToRHSLiteral (r@(RHSLiteral True _), index) =
...
+     aggVar = GroupVar rule rhsidx
+     group_var = "__debug_" ++ rhsVar r
+     input1 = head $ Compile.recordAfterPrefix d rule (rhsidx - 1)
+     input = eTuple [ input1, rhsProject r ]
+     rAgg = RHSGroupBy { rhsVar     = group_var
+                       , rhsProject = input
+                       , rhsGroupBy = rhsGroupBy r
+                       }
+     tinputs = tOpaque (mOD_STD ++ "::Vec") [exprType d (CtxRuleRProject rule rhsidx) input1]
+     rCond = RHSCondition { rhsExpr = eSet (eTuple [eVarDecl "__inputs" tinputs, eVarDecl (rhsVar r) (varType d aggVar)])
```
[the remark this answers is not in the document]
Comment (ryzhyk) [reply src/Language/DifferentialDatalog/Debug.hs:79]: I assume that valid user-defined names cannot be prefixed by two underscores (but we have to actually start enforcing this).

File: src/Language/DifferentialDatalog/Syntax.hs (hunk 3 above)
Comment (ryzhyk) [reply src/Language/DifferentialDatalog/Syntax.hs:597]: Nope.  But I cannot think of a better name. Filter is probably less confusing.

File: test/datalog_tests/group_test.dl (hunk 4 above)
Comment (ryzhyk) [reply test/datalog_tests/group_test.dl:76]: Yes, I am forcing DDlog to create an intermediate relation and store the group in it to test this functionality.

**Pull request merged**